src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs

```diff
@@ -357,6 +359,107 @@ public static class Extensions
             return CS8VerifyAnalyzerAsync(originalCode);
         }
 
+        [Fact]
+        [WorkItem(3786, "https://github.com/dotnet/roslyn-analyzers/issues/3786")]
+        public Task CS_NoDiagnostic_ParametersDifferMoreThanOne()
+        {
+            return CS8VerifyAnalyzerAsync(@"
+using System;
+using System.Threading;
+class C
+{
+    void MyMethod(int i) {}
+    void MyMethod(int i, bool b) {}
+    void MyMethod(int i, bool b, CancellationToken c) {}
+
+    public void M(CancellationToken ct)
+    {
+        MyMethod(1);
+    }
+}
+            ");
+        }
+
+        [Fact]
+        [WorkItem(3786, "https://github.com/dotnet/roslyn-analyzers/issues/3786")]
+        public Task CS_NoDiagnostic_LambdaAndExtensionMethod_NoTokenInLambda()
+        {
+            // Only for local methods will we look for the ct in the top-most ancestor
+            // For anonymous methods we will only look in the immediate ancestor
+            return VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Threading;
+public static class Extensions
+{
+    public static void Extension(this bool b, Action<int> action) {}
+    public static void MyMethod(this int i, CancellationToken c = default) {}
+}
+class C
+{
+    public void M(CancellationToken ct)
+    {
+        bool b = false;
+        b.Extension((j) =>
+        {
+            j.MyMethod();
+        });
+    }
+}
+            ");
+        }
+
+        [Fact]
+        [WorkItem(3786, "https://github.com/dotnet/roslyn-analyzers/issues/3786")]
+        public Task CS_NoDiagnostic_AnonymousDelegateAndExtensionMethod_NoTokenInAnonymousDelegate()
+        {
+            // Only for local methods will we look for the ct in the top-most ancestor
+            // For anonymous methods we will only look in the immediate ancestor
+            return VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Threading;
+public static class Extensions
+{
+    public delegate void MyDelegate(int i);
+    public static void Extension(this bool b, MyDelegate d) {}
+    public static void MyMethod(this int i, CancellationToken c = default) {}
+}
+class C
+{
+    public void M(CancellationToken ct)
+    {
+        bool b = false;
+        b.Extension((int j) =>
+        {
+            j.MyMethod();
+        });
+    }
+}
+            ");
+        }
+
+        [Fact]
+        [WorkItem(3786, "https://github.com/dotnet/roslyn-analyzers/issues/3786")]
+        public Task CS_NoDiagnostic_StaticLocalMethod()
+        {
+            // Local static functions are available in C# >= 8.0
+            return CS8VerifyAnalyzerAsync(@"
+using System;
+using System.Threading;
+class C
+{
+    public static void MyMethod(int i, CancellationToken c = default) {}
+    public void M(CancellationToken ct)
+    {
+        LocalStaticMethod();
+        static void LocalStaticMethod()
+        {
+            MyMethod(5);
```

Comment:
This should generate a diagnostic. We should not be excluding these.

Reply:
Please also add a test where immediate outer local function is not static but its container is a static local function. We should generate a diagnostic even for that case:

```csharp
public static void MyMethod(int i, CancellationToken c = default) {}
    public void M(CancellationToken ct)
    {
        LocalStaticMethod();
        static void LocalStaticMethod()
        {
            LocalMethod();
            void LocalMethod()
            {
                MyMethod(5); // This should generate a diagnostic, but no code fix
            }
        }
}
```

Reply:
Basically, refactoring a large method to have local functions (nested or non-nested ones, static or non-static ones) should not alter the diagnostics reported by this analyzer. Local functions are primarily used to refactor code for readability, unlike lambdas which are passed to other methods and need to ensure no captures to avoid allocations.

Reply:
@mavasani Can you explain why having a static local function should trigger a diagnostic? According to the [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/static-local-functions), the token would not not reachable inside the static local method:

> _"A local function declared static cannot capture state from the enclosing scope. As a result, locals, parameters, and this from the enclosing scope are not available within a static local function."._

I tested the above code in a console app, and I get errors if I attempt to consume the token in any of the two cases:

![image](https://user-images.githubusercontent.com/1175054/86043669-e2228900-b9fd-11ea-8c1c-05d490cc149a.png)

My understanding was that suggesting a diagnostic without offering a fix was for cases that were valid but the fixer had limitations preventing it from offering a fix. In this case, we would be offering a diagnostic for a case that is unsupported/invalid.

Can you please tell me if I'm misunderstanding?

I can add the additional unit test you requested, but I would like to clarify this first.

Reply:
> the token would not not reachable inside the static local method:

Yes, I don't expect a code fix for this case. User would need to either convert it to a non-static local function OR thread in a new cancellation token parameter to the static local function. Both of these are absolutely acceptable changes given it leads to correctly passing in a cancellation token to a cancellable API invoked from the static local method.

Reply:
I see. Thanks for the explanation.
I will update the analyzer so it offers a diagnostic, and will work on ensuring the fixer bails out when it detects this case.

Reply:
Addressed, @mavasani. PTAL whenever you get a chance.